File: 14-cryptography/04-bigInt/extEuclid.py
def ext_euclid(a, b):
    old_s, s = 1, 0
    old_t, t = 0, 1
    old_r, r = a, b
    if b == 0:
        return 1, 0, a
    else:
        while(r!=0):
            q = old_r // r
            old_r, r = r, old_r-q*r
            old_s, s = s, old_s-q*s
            old_t, t = t, old_t-q*t
    return old_s, old_t, old_r

print('ext_euclid(30,27)=', ext_euclid(30,27))

# $ python extEuclid.py
# ext_euclid(30,27)= (1, -1, 3)

File: 14-cryptography/04-bigInt/extEuclid.js
// gcd(a,b)= ri =si+bti
export function extEuclid(a, b) {
    // var s0, s, t, t0, r, r0
    let [s0,s] = [1n,0n] // let s0 = 1n, s = 0n // let [s0,s] = [1n,0n]
    let [t0,t] = [0n,1n] // let t0 = 0n, t = 1n // let [t0,t] = [0n,1n]
    let [r0,r] = [a,b] // let r0 = a, r = b   // let [r0,r] = [a,b]
    if (b === 0n)
        return [1n, 0n, a]
    else {
        while (r != 0n) {
            let q = r0/r
            let rt=r; r=r0-q*rt; r0=rt; // [r0, r] = [r, r0-q*r]
            let st=s; s=s0-q*st; s0=st; // [s0, s] = [s, s0-q*s]
            let tt=t; t=t0-q*tt; t0=tt; // [t0, t] = [t, t0-q*t]
        }
    }
    return [s0, t0, r0]
}

console.log('extEuclid(30n, 27n)=', extEuclid(30n, 27n))
